// regexpress/dist/regexbuilder/regex_assertion_builder.js
import { stringOrRegExp } from "../utils/string_or_regexp.js";
import { groupStarters } from "./process_group_code.js";
import { RegexBuilderBase } from "./regex_builder_base.js";
export class RegexAssertionBuilder extends RegexBuilderBase {
    /** Adds a word boundary to the regex. */
    boundary() {
        this.regex.parts.push('\\b');
        return this;
    }
    /** Adds a word boundary to the regex. */
    negatedBoundary() {
        this.regex.parts.push('\\B');
        return this;
    }
    /** Adds the provided content to the regex within word boundaries `\bcontent\b` */
    bordered(content) {
        this.regex.parts.push(`\\b${content}\\b`);
        return this;
    }
    /** Adds a caret character, or start-of-the-line assertion, to the regex. */
    lineStart() {
        this.regex.parts.push('^');
        return this;
    }
    /** Adds a dollar character, or end-of-the-line assertion, to the regex. */
    lineEnd() {
        this.regex.parts.push('$');
        return this;
    }
    /** Combines RegexBuilder methods `lineStart()` and `add()`. */
    startsWith(start) {
        this.regex.parts.push('^', stringOrRegExp(start));
        return this;
    }
    /** Combines RegexBuilder methods add(...) and lineEnd(). */
    endsWith(end) {
        this.regex.parts.push(end, '$');
        return this;
    }
    /** Adds a lookahead assertion to the regex. */
    lookahead(la) {
        this.regex.parts.push(groupStarters.la, stringOrRegExp(la), groupStarters.close);
        return this;
    }
    /** Adds a lookbehind assertion  to the regex. */
    lookbehind(lb) {
        this.regex.parts.push(groupStarters.lb, stringOrRegExp(lb), groupStarters.close);
        return this;
    }
    /** Adds a negated lookahead assertion  to the regex. */
    negatedLA(nla) {
        this.regex.parts.push(groupStarters.nla, stringOrRegExp(nla), groupStarters.close);
        return this;
    }
    /** Adds a negated lookbehind assertion to the regex. */
    negatedLB(nlb) {
        this.regex.parts.push(groupStarters.nlb, stringOrRegExp(nlb), groupStarters.close);
        return this;
    }
    /** Alias for `lookahead()`. */
    followedBy(la) {
        return this.lookahead(la);
    }
    /** Alias for `negatedLA()`. */
    notFollowedBy(nla) {
        return this.negatedLA(nla);
    }
    /** Alias for `lookbehind()`. */
    precededBy(lb) {
        return this.lookbehind(lb);
    }
    /** Alias for `negatedLB()`. */
    notPrecededBy(nlb) {
        return this.negatedLB(nlb);
    }
}
